nvflare/private/fed/server/server_runner.py

```diff
@@ -268,33 +269,14 @@ def process_task_request(self, client: Client, fl_ctx: FLContext) -> (str, str,
             self.log_debug(fl_ctx, "firing event EventType.BEFORE_TASK_DATA_FILTER")
             self.fire_event(EventType.BEFORE_TASK_DATA_FILTER, fl_ctx)
 
-            # apply scope filters first
-            scope_object = fl_ctx.get_prop(FLContextKey.SCOPE_OBJECT)
-            filter_list = []
-            if scope_object:
-                assert isinstance(scope_object, Scope)
-                if scope_object.task_data_filters:
-                    filter_list.extend(scope_object.task_data_filters)
+            task_filter_list = self.config.task_data_filters.get(task_name + FilterChain.DELIMITER + FilterChain.OUT)
+            filter_error, task_data = apply_data_filters(task_filter_list, task_data, self.logger, fl_ctx)
 
-            task_filter_list = self.config.task_data_filters.get(task_name)
-            if task_filter_list:
-                filter_list.extend(task_filter_list)
-
-            if filter_list:
-                for f in filter_list:
-                    try:
-                        task_data = f.process(task_data, fl_ctx)
-                    except Exception as e:
-                        self.log_exception(
-                            fl_ctx,
-                            "processing error in task data filter {}: {}; "
-                            "asked client to try again later".format(type(f), secure_format_exception(e)),
-                        )
-
-                        with self.wf_lock:
-                            if self.current_wf:
-                                self.current_wf.responder.handle_exception(task_id, fl_ctx)
-                        return self._task_try_again()
+            if filter_error:
+                with self.wf_lock:
+                    if self.current_wf:
+                        self.current_wf.responder.handle_exception(task_id, fl_ctx)
+                return self._task_try_again()
```

Comment:
why is this not something like `return make_reply(ReturnCode.TASK_DATA_FILTER_ERROR)`?

Reply:
This is in the process chain of getting task for client. when it fails, it will skip the current request and asks the client to try_again.

Reply:
What if there IS really a problem with `task_data_filters`?
Like implementation errors by the users.
I think we should warn the users, otherwise we will have clients keep trying again.

Reply:
Good point. Maybe we should log the error and then end the job?

Reply:
This has been our original server runner behavior from the beginning. The idea is not "fail hard". We did log the exception to notify the users. I could see the reasons not to end the job run as well. Not sure if we should change this behavior.